### PR TITLE
HTTP/3: Complete support for UseHttps

### LINKED ISSUE #34858: HTTP/3: UseHttps configuration
Ensure that we set any options configured with UseHttps with HTTP/3. Some values might not be supported by QUIC.

### LINKED ISSUE #42831: Add TlsConnectionOptions to connection abstractions
## Background and Motivation

Kestrel needs to pass TLS configuration to the QUIC transport. Neither project has a dependency on the other.

Communication is done through values added to IFeatureCollection. Rather than add a loose collection of types and delegates to the feature collection, I'd rather have a strongly typed options type that has the necessary configuration.

Also, part of TLS configuration is a callback that's called for each connection to resolve `SslServerAuthenticationOptions`. If the callback takes a context object, then new values can be added in the future.

## Proposed API

Note: Microsoft.AspNetCore.Connections.Abstractions targets .NET Framework and .NET Standard. These types would only be present in .NET 7 it's the only target that supports `SslServerAuthenticationOptions` 

```cs
namespace Microsoft.AspNetCore.Connections;

/// <summary>
/// Options used to configure a per connection callback for TLS configuration.
/// </summary>
public class TlsConnectionCallbackOptions
{
    /// <summary>
    /// The callback to invoke per connection. This property is required.
    /// </summary>
    public Func<TlsConnectionCallbackContext, ValueTask<SslServerAuthenticationOptions>> OnConnection { get; set; } = default!;

    /// <summary>
    /// Optional application state to flow to the <see cref="OnConnection"/> callback.
    /// </summary>
    public object? OnConnectionState { get; set; }

    /// <summary>
    /// Gets or sets a list of ALPN protocols.
    /// </summary>
    public List<SslApplicationProtocol> ApplicationProtocols { get; set; } = default!;
}

/// <summary>
/// Per connection state used to determine the TLS options.
/// </summary>
public class TlsConnectionCallbackContext
{
    /// <summary>
    /// Information from the Client Hello message.
    /// </summary>
    public SslClientHelloInfo ClientHelloInfo { get; set; }

    /// <summary>
    /// The information that was passed when registering the callback.
    /// </summary>
    public object? State { get; set; }

    /// <summary>
    /// The token to monitor for cancellation requests.
    /// </summary>
    public CancellationToken CancellationToken { get; set; }

    /// <summary>
    /// Information about an individual connection.
    /// </summary>
    public ConnectionContext Connection { get; set; } = default!;
}
```

## Usage Examples

```csharp
var features = new FeaturesCollection();

features.Set(new TlsConnectionCallbackOptions
{
    ApplicationProtocols = new List<SslApplicationProtocol> { SslApplicationProtocol.Http3 },
    OnConnection = context =>
    {
        return listenOptions.HttpsCallbackOptions.OnConnection(new TlsHandshakeCallbackContext
        {
            ClientHelloInfo = context.ClientHelloInfo,
            CancellationToken = context.CancellationToken,
            State = context.State,
            Connection = context.Connection,
        });
    },
    OnConnectionState = listenOptions.HttpsCallbackOptions.OnConnectionState,
});
```

## Alternative Designs

These types are very similar to what is in Kestrel:
* https://github.com/dotnet/aspnetcore/blob/0602a3266ea9d116df479d558cd99e854d4bc03d/src/Servers/Kestrel/Core/src/TlsHandshakeCallbackOptions.cs
* https://github.com/dotnet/aspnetcore/blob/0602a3266ea9d116df479d558cd99e854d4bc03d/src/Servers/Kestrel/Core/src/TlsHandshakeCallbackContext.cs

### PR DESCRIPTION
**Pull request by JamesNK**
Fixes https://github.com/dotnet/aspnetcore/issues/42831
Fixes https://github.com/dotnet/aspnetcore/issues/34858

Question:
* We need to pass TLS configuration information from Kestrel to Quic transport. The PR is setting a delegate that returns an `SslServerAuthenticationOptions` and an `IList<SslApplicationProtocol>` on a feature collection. Should a new type be added to the connection abstractions project for sharing this information? Delegates and lists in a feature collection feels a bit hacky. e.g. https://github.com/dotnet/aspnetcore/blob/06a1f45f79c300c732b0171ce6318a130d05a083/src/Servers/Kestrel/Core/src/TlsHandshakeCallbackOptions.cs
* Should the delegate take a context object in case we need to add new values in the future? e.g. https://github.com/dotnet/aspnetcore/blob/06a1f45f79c300c732b0171ce6318a130d05a083/src/Servers/Kestrel/Core/src/TlsHandshakeCallbackContext.cs

(can't use these `TlsHandshakeCallbackOptions` or `TlsHandshakeCallbackContext` because they're in Kestrel and Quic transport doesn't have a dependency)

### CODE REVIEW

File: src/Servers/Kestrel/Transport.Quic/src/Internal/QuicConnectionListener.cs [hunk 1]
```diff
@@ -33,30 +39,38 @@ public QuicConnectionListener(QuicTransportOptions options, ILogger log, EndPoin
...
             ListenBacklog = options.Backlog,
-            ConnectionOptionsCallback = (connection, helloInfo, cancellationToken) =>
+            ConnectionOptionsCallback = async (connection, helloInfo, cancellationToken) =>
             {
+                var serverAuthenticationOptions = await sslServerAuthenticationOptionsCallback(helloInfo, cancellationToken);
+
+                // If the SslServerAuthenticationOptions doesn't have a cert or protocols then the
+                // QUIC connection will fail and the client receives an unhelpful message.
+                // Validate the options on the server and log issues to improve debugging.
+                ValidateServerAuthenticationOptions(serverAuthenticationOptions);
```
Comment (Tratcher) [inline src/Servers/Kestrel/Transport.Quic/src/Internal/QuicConnectionListener.cs:62]: In another code path we set APLN for them if it wasn't set in the callback:
https://github.com/dotnet/aspnetcore/blob/5834214ddeab320106c8e728b92b611a8c6add63/src/Servers/Kestrel/Core/src/Middleware/HttpsConnectionMiddleware.cs#L431-L437

File: src/Servers/Kestrel/Transport.Quic/src/Internal/QuicConnectionListener.cs [hunk 2]
```diff
@@ -65,6 +79,20 @@ public QuicConnectionListener(QuicTransportOptions options, ILogger log, EndPoin
         EndPoint = listenEndPoint;
     }
 
+    private void ValidateServerAuthenticationOptions(SslServerAuthenticationOptions serverAuthenticationOptions)
```
[the remark this answers is not in the document]
Comment (JamesNK) [reply src/Servers/Kestrel/Transport.Quic/src/Internal/QuicConnectionListener.cs:82]: This is in the `QuicListenerOptions.ConnectionOptionsCallback`. Errors will get eated.

File: src/Servers/Connections.Abstractions/src/PublicAPI.Unshipped.txt [hunk 3]
```diff
@@ -1 +1,19 @@
 #nullable enable
+Microsoft.AspNetCore.Connections.TlsConnectionCallbackContext
```
Comment (JamesNK) [inline src/Servers/Connections.Abstractions/src/PublicAPI.Unshipped.txt:2]: I added some types to Connections.Abstractions but they use SslServerAuthenticationOptions which isn't available in .NET Standard or .NET Framework. I put an ifdef around them for .NET 7 or later only, but I get:

> src/Servers/Connections.Abstractions/src/PublicAPI.Unshipped.txt(2,1): error RS0017: (NETCORE_ENGINEERING_TELEMETRY=Build) Symbol 'Microsoft.AspNetCore.Connections.TlsConnectionCallbackContext' is part of the declared API, but is either not public or could not be found

How can I fix this?

Comment (JamesNK) [reply src/Servers/Connections.Abstractions/src/PublicAPI.Unshipped.txt:2]: cc @dougbu @wtgodbe

Comment (BrennanConroy) [reply src/Servers/Connections.Abstractions/src/PublicAPI.Unshipped.txt:2]: https://github.com/dotnet/roslyn-analyzers/blob/main/src/PublicApiAnalyzers/PublicApiAnalyzers.Help.md#conditional-api-differences

Comment (dougbu) [reply src/Servers/Connections.Abstractions/src/PublicAPI.Unshipped.txt:2]: See also what @davidfowl added in the Microsoft.AspNetCore.SignalR.Common.csproj project and its src/SignalR/common/SignalR.Common/src/PublicAPI folder

File: src/Servers/Kestrel/Core/src/Internal/Infrastructure/TransportManager.cs [hunk 4]
```diff
@@ -57,14 +60,81 @@ public async Task<EndPoint> BindAsync(EndPoint endPoint, MultiplexedConnectionDe
...
-            features.Set(HttpsConnectionMiddleware.CreateHttp3Options(listenOptions.HttpsOptions));
+            var sslServerAuthenticationOptions = HttpsConnectionMiddleware.CreateHttp3Options(listenOptions.HttpsOptions);
+            features.Set(new TlsConnectionCallbackOptions
+            {
+                ApplicationProtocols = sslServerAuthenticationOptions.ApplicationProtocols ?? new List<SslApplicationProtocol> { SslApplicationProtocol.Http3 },
+                OnConnection = (context, cancellationToken) => ValueTask.FromResult(sslServerAuthenticationOptions),
+                OnConnectionState = null,
+            });
+        }
+        if (listenOptions.HttpsCallbackOptions != null)
```
Comment (adityamandaleeka) [inline src/Servers/Kestrel/Core/src/Internal/Infrastructure/TransportManager.cs:71]: else if?

File: src/Servers/Kestrel/Core/src/Internal/Infrastructure/TransportManager.cs [hunk 5]
```diff
@@ -57,14 +60,81 @@ public async Task<EndPoint> BindAsync(EndPoint endPoint, MultiplexedConnectionDe
         // The transport will check if the feature is missing.
```
Comment (adityamandaleeka) [inline src/Servers/Kestrel/Core/src/Internal/Infrastructure/TransportManager.cs:60]: Does "This should always be set in production" need to be clarified?

Comment (JamesNK) [reply src/Servers/Kestrel/Core/src/Internal/Infrastructure/TransportManager.cs]: I've updated the comment to be clearer.

File: src/Servers/Connections.Abstractions/src/Microsoft.AspNetCore.Connections.Abstractions.csproj [hunk 6]
```diff
@@ -19,10 +19,13 @@
...
+  <ItemGroup>
+    <AdditionalFiles Include="PublicAPI/$(TargetFramework)/PublicAPI.Shipped.txt" />
+    <AdditionalFiles Include="PublicAPI/$(TargetFramework)/PublicAPI.Unshipped.txt" />
+  </ItemGroup>
+
   <!-- Special case building from source because Microsoft.Bcl.AsyncInterfaces isn't available for source builds. -->
-  <ItemGroup Condition="'$(TargetFramework)' == 'netstandard2.0' OR
-      '$(TargetFramework)' == '$(DefaultNetFxTargetFramework)' OR
-      ('$(MSBuildRestoreSessionId)' == '' AND '$(DotNetBuildFromSource)' != 'true') ">
+  <ItemGroup Condition="'$(TargetFramework)' == 'netstandard2.0' OR&#xD;&#xA;      '$(TargetFramework)' == '$(DefaultNetFxTargetFramework)' OR&#xD;&#xA;      ('$(MSBuildRestoreSessionId)' == '' AND '$(DotNetBuildFromSource)' != 'true') ">
```
Comment (wtgodbe) [inline src/Servers/Connections.Abstractions/src/Microsoft.AspNetCore.Connections.Abstractions.csproj:28]: This looks like it can be undone

File: src/Servers/Kestrel/Kestrel.slnf [hunk 7]
```diff
@@ -2,11 +2,17 @@
   "solution": {
     "path": "..\\..\\..\\AspNetCore.sln",
     "projects": [
+      "src\\DataProtection\\Abstractions\\src\\Microsoft.AspNetCore.DataProtection.Abstractions.csproj",
```
Comment (adityamandaleeka) [inline src/Servers/Kestrel/Kestrel.slnf:5]: Are these changes required/intentional?

Comment (JamesNK) [reply src/Servers/Kestrel/Kestrel.slnf:5]: They're required to build the Kestrel solution. The recently added web transport sample project uses WebApplication and references the `Microsoft.AspNetCore` project. It requires these dependencies.

**Review by adityamandaleeka — APPROVED** [review]

**Comment by adityamandaleeka** [comment]
@JamesNK I put the blog-candidate label on this PR, but we can probably do a section including this and the other HTTP/3 changes going into RC 1.
